File: internal/auth/pkce.go
package auth

import (
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"errors"
	"regexp"
)

// pkceS256ChallengePattern は RFC 7636 §4.2 の S256 code_challenge 形式。
// challenge = BASE64URL-ENCODE(SHA256(verifier)) は padding なし 43 文字に固定される。
var pkceS256ChallengePattern = regexp.MustCompile(`^[A-Za-z0-9_-]{43}$`)

// pkceS256VerifierPattern は RFC 7636 §4.1 の code_verifier 形式。
// unreserved 文字（ALPHA / DIGIT / "-" / "." / "_" / "~"）で構成され、長さは 43〜128 文字。
// 本実装は RFC が定義する文字集合に厳密準拠する。
var pkceS256VerifierPattern = regexp.MustCompile(`^[A-Za-z0-9._~-]{43,128}$`)

// ValidatePKCES256 は flow=native の PKCE パラメータを検証する。
//
// method は "S256" の厳密一致のみ受理し（plain・小文字・欠落は拒否、NFR 1.2）、
// challenge は RFC 7636 §4.2 の S256 形式（base64url no-padding、43 文字）のみ受理する。
// 合格時は nil、不合格時は理由を表す error を返す。呼び出し側は error 文字列を
// クライアントへ反射せず、固定メッセージで応答すること（NFR 1.3）。
func ValidatePKCES256(challenge, method string) error {
	if method != "S256" {
		return errors.New("pkce method must be S256")
	}
	if challenge == "" {
		return errors.New("pkce challenge is required")
	}
	if !pkceS256ChallengePattern.MatchString(challenge) {
		return errors.New("pkce challenge is not a valid S256 challenge")
	}
	return nil
}

// VerifyPKCES256Verifier は POST /api/auth/token の交換時に提示された code_verifier が、
// auth_code 発行時に保存された storedChallenge（S256 challenge）と一致するかを定数時間で検証する。
//
// RFC 7636 §4.1 の verifier 形式（unreserved 43〜128 文字）を満たさない場合は false。
// 形式が妥当な場合は SHA-256 を base64url（no-padding）でエンコードした派生値を求め、
// storedChallenge と crypto/subtle.ConstantTimeCompare で比較する（NFR 1.4: タイミング攻撃回避）。
//
// 戻り値は (一致するか否か) の bool のみで、理由を区別する情報は返さない（Req 2.6 の応答
// uniform 化に対応）。本関数は平文 verifier をログに出さないこと（NFR 1.3 は呼び出し側でも守る）。
func VerifyPKCES256Verifier(verifier, storedChallenge string) bool {
	// 形式チェックを最初に行うことで、明らかに不正な入力に対して SHA-256 計算を回避する。
	if !pkceS256VerifierPattern.MatchString(verifier) {
		return false
	}
	if storedChallenge == "" {
		return false
	}

	sum := sha256.Sum256([]byte(verifier))
	derived := base64.RawURLEncoding.EncodeToString(sum[:])

	return subtle.ConstantTimeCompare([]byte(derived), []byte(storedChallenge)) == 1
}

File: web/src/lib/pkce.ts
/**
 * PKCE (RFC 7636) の code_verifier / code_challenge (S256) をブラウザ WebCrypto で生成する
 * 純粋 utility。
 *
 * サーバ側（`internal/auth/pkce.go` の `VerifyPKCES256Verifier`）は
 * `sha256.Sum256([]byte(verifier))` を計算する。ここで入力される `verifier` は
 * code_verifier "文字列" そのものである。本モジュールは対称に、code_verifier 文字列
 * （43 文字 base64url）を `TextEncoder` でエンコードしたバイト列を SHA-256 に入力し、
 * その結果を base64url 化して code_challenge を得る。生成された 32 バイト crypto
 * random を直接 SHA-256 に食わせるとサーバと一致しないため注意。
 *
 * 副作用なし・throw なし（WebCrypto API 由来の rejection はそのまま呼び出し側に伝播）。
 * 生成値は本モジュール内で保持せず、返却値としてのみ渡す（NFR 1.1）。
 */

/** PKCE code_verifier と code_challenge (S256) の組。 */
export interface PkcePair {
  /** 32 バイト crypto random を base64url 化した 43 文字文字列（無 padding） */
  codeVerifier: string;
  /** SHA-256(code_verifier as UTF-8 bytes) を base64url 化した 43 文字文字列（無 padding） */
  codeChallenge: string;
}

/**
 * Uint8Array を base64url 文字列（`+/=` を含まない）にエンコードする内部ヘルパ。
 *
 * `btoa` は Latin-1 のみ受け付けるため、まず `String.fromCharCode` で byte-per-char の
 * バイナリ文字列に組み直してから base64 化し、base64 → base64url 差分（`+`→`-`,
 * `/`→`_`, 末尾 `=` 除去）を適用する。
 */
function bytesToBase64url(bytes: Uint8Array): string {
  let binary = "";
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  const b64 = btoa(binary);
  return b64.replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

/**
 * code_verifier 文字列から PKCE code_challenge (S256) を導出する純粋関数。
 *
 * サーバ側 `internal/auth/pkce.go` の `sha256.Sum256([]byte(verifier))` に対称の入力を
 * 与えるため、`TextEncoder.encode(codeVerifier)`（= UTF-8/ASCII バイト列）を
 * `crypto.subtle.digest("SHA-256", ...)` に渡す。同関数を独立に export しておくことで
 * RFC 7636 Appendix B の既知ベクトル（固定 verifier → 期待 challenge）で SHA-256 派生
 * の正しさを検証できる（`generatePkcePair` は乱数入力のため直接には検証不能）。
 *
 * @param codeVerifier 43〜128 文字の RFC 7636 §4.1 code_verifier 文字列
 * @returns S256 派生の base64url 化 code_challenge（43 文字、無 padding）
 */
export async function deriveCodeChallengeS256(codeVerifier: string): Promise<string> {
  const verifierBytes = new TextEncoder().encode(codeVerifier);
  const digest = await crypto.subtle.digest("SHA-256", verifierBytes);
  return bytesToBase64url(new Uint8Array(digest));
}

/**
 * PKCE code_verifier / code_challenge (S256) をブラウザ WebCrypto で生成する。
 *
 * - code_verifier: `crypto.getRandomValues(new Uint8Array(32))` の base64url 化（43 文字）
 * - code_challenge: `deriveCodeChallengeS256(codeVerifier)`（= サーバ検証と対称の入力を SHA-256）
 *
 * 生成値は本関数の返却値としてのみ渡され、モジュール内に保持しない（NFR 1.1）。
 * WebCrypto API 側の rejection は握り潰さず呼び出し側に伝播する。
 */
export async function generatePkcePair(): Promise<PkcePair> {
  const randomBytes = new Uint8Array(32);
  crypto.getRandomValues(randomBytes);
  const codeVerifier = bytesToBase64url(randomBytes);
  const codeChallenge = await deriveCodeChallengeS256(codeVerifier);
  return { codeVerifier, codeChallenge };
}
